README.md

```diff
@@ -511,6 +506,33 @@ False
 ([3, 1, 2], PyTreeSpec(OrderedDict([('b', [*]), ('a', [*, *])])))
 ```
 
+Since OpTree v0.9.0, the `tree_map` and `tree_map_with_path` functions will preserve the original key ordering in the output pytree.
+
+```python
+>>> leaves, treespec = optree.tree_flatten({'b': [3], 'a': [1, 2]})
+>>> optree.tree_unflatten(treespec, leaves)
+{'a': [1, 2], 'b': [3]}
```

Comment:
This looks confusing to me and is an unnecessary limitation. I'm pretty sure this will lead to bugs.

For example, `dm-tree` or `tf.nest` both keep insertion order without this limitation:

```python
d = {'b': [3], 'a': [1, 2]}

leaves = tree.flatten(d)
d2 = tree.unflatten_as(d, leaves)

assert list(d2.keys()) == list(d.keys())  # Dict key order preserved
```

As explained in https://github.com/google/jax/issues/4085#issuecomment-1186908662

* During flattening, all keys are sorted (so all dict can be comparable)
* The `treespec` preserve `dict` key order (like a regular dict). When unflattening the dict, the keys are re-ordered to match the structure.

For a reference Python implementation: https://github.com/google/etils/blob/9ab6c675daabbed84fd9c97d337fa3a3abe76e7a/etils/etree/backend.py#LL212C7-L212C65 (see the `sorted(structure.items())` during the first iter, but not in the second one).

```python
import etils.etree
etree = etree.py.backend

leaves, structure = etree.flatten({'b': [3], 'a': [1, 2]})
print(leaves)  # [1, 2, 3]
d2 = etree.unflatten(structure, leaves)
print(d2)  # {'b': [3], 'a': [1, 2]}  << KEY PRESERVED !!
```

Reply:
> This looks confusing to me and is an unnecessary limitation. I'm pretty sure this will lead to bugs.

@Conchylicultor Thanks for the feedback. I agree with this. Keep the insertion order is already implemented in this PR. I wonder if should we enable it as the default behavior. Enable it for tree-map introduces approx 25% performance drop for `builtins.dict` and `collections.defaultdict`. cc @JieRen98 @Benjamin-eecs

One reason for keeping the sorted order in unflatten output is to keep the referential transparency in serialization. For example, you can serialize the `treespec` instance and deserialize it in another process.

```python
import multiprocessing as mp

# process 1
leaves, treespec = optree.tree_flatten(tree)

proc = mp.Process(target=func, args=(treespec,))  # start a new process (process 2)
```

Should process 2 memorize the original key order as it is in process 1?

If we want this, serialization/deserialization should not erase the memory on the input dict keys. I have not added this yet, but it would be easy. Adding new fields in the pickler may be a breaking change. But this is not a strict blocker.

------

> ```python
> d = {'b': [3], 'a': [1, 2]}
> 
> leaves = tree.flatten(d)
> d2 = tree.unflatten_as(d, leaves)
> 
> assert list(d2.keys()) == list(d.keys())  # Dict key order preserved
> ```

dm-tree does not have a concept of "treespec" or "treedef". It iterates over the referenced tree object as the output structure.

```python
d = {'b': [3], 'a': [1, 2]}
d_ref = {'a': [1, 2], 'b': [3]}

leaves = tree.flatten(d)
d2 = tree.unflatten_as(d_ref, leaves)

assert list(d2.keys()) != list(d.keys())      # Dict key order "not preserved"
assert list(d2.keys()) == list(d_ref.keys())  # Dict key order is same as the referenced object
```

Also, if we memorize the input key order in `treespec`, the output pytree is also depending on the "referenced treespec".

```python
In [1]: import optree

In [2]: d = {'b': [3], 'a': [1, 2]}

In [3]: leaves, treespec = optree._C.flatten(d, None, False, '', unflatten_preserve_dict_order=True)

In [4]: treespec_ref = optree._C.flatten({'a': [1, 2], 'b': [3]}, None, False, '', unflatten_preserve_dict_order=True)[1]

In [5]: treespec
Out[5]: PyTreeSpec({'a': [*, *], 'b': [*]})

In [6]: treespec_ref
Out[6]: PyTreeSpec({'a': [*, *], 'b': [*]})

In [7]: treespec == treespec_ref
Out[7]: True

In [8]: treespec.unflatten(leaves)
Out[8]: {'b': [3], 'a': [1, 2]}

In [9]: treespec_ref.unflatten(leaves)
Out[9]: {'a': [1, 2], 'b': [3]}

In [10]: list(treespec.unflatten(leaves)) == list(treespec_ref.unflatten(leaves))
Out[10]: False
```

> For a reference Python implementation: https://github.com/google/etils/blob/9ab6c675daabbed84fd9c97d337fa3a3abe76e7a/etils/etree/backend.py#LL212C7-L212C65 (see the `sorted(structure.items())` during the first iter, but not in the second one).
> 
> ```python
> import etils.etree
> etree = etils.etree.py.backend
> 
> leaves, structure = etree.flatten({'b': [3], 'a': [1, 2]})
> print(leaves)  # [1, 2, 3]
> d2 = etree.unflatten(structure, leaves)
> print(d2)  # {'b': [3], 'a': [1, 2]}  << KEY PRESERVED !!
> ```

In your example of `etils.etree`, the returned `structure` is the input `tree` itself. That explains why `etils.etree` preserves key order. The referenced pytree is the input pytree.

```python
In [1]: import etils.etree
   ...: etree = etils.etree.py.backend

In [2]: tree = {'b': [3], 'a': [1, 2]}

In [3]: leaves, structure = etree.flatten(tree)

In [4]: structure is tree
Out[4]: True

In [5]: structure
Out[5]: {'b': [3], 'a': [1, 2]}

In [6]: type(structure)
Out[6]: <class 'dict'>
```

Reply:
I have enabled the `unflatten_preserve_dict_order` flag globally. Waiting for response from other team members.

Reply:
I think it is necessary to preserve dict key order in the output, even if it will lose some performance because this is a necessary thing to do, not a thing that needs to be weighed against performance loss.